### ISSUE TITLE
ld: library not found for -lgcc_s.10.5 (macOS 10.14.3)

### ISSUE THREAD

**Issue by megrimm** [pure-data/pd-lib-builder#49]
getting this error with pd-lib-builder 5.1

```
ld: library not found for -lgcc_s.10.5
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

I do NOT get this error with pd-lib-builder 5.0 meaning everything compiles just fine.
I tried running a diff on the two but fail to see what has changed to cause error. Maybe im missing a new flag I need to add in my Makefile?

**Comment by katjav** [comment]
This issue may well be introduced with commit ece8e6b, which intended to fix a bug where minimum OSX version wasn't passed at all. The flag is needed to make sure that the object will support older OSX versions. But apparently, OSX 10.14.3 doesn't support building for 10.5.

Can you check what happens when changing '10.5' to '10.6' in line 581 (I don't have 10.14 or any recent OSX to test).

**Comment by megrimm** [comment]
>> changing '10.5' to '10.6' in line 58

yup that was it... works for me now with that change. thanks!

**Comment by katjav** [comment]
Good, thanks for testing and reporting. I'll update pdlibbuilder accordingly.

**Comment by katjav** [comment]
Should be fixed with commit 1137cca.

**Comment by katjav** [comment]
@megrimm, now that we're at it could you tell me if your OSX can still build for i386? By default pdlibbuilder tries to build a fat binary, and I guess you would get errors if it can't, but I don't know if you use that default.

**Comment by megrimm** [comment]
no error. looks like it does build for i386? I do get a depreciation warning:

`ld: warning: The i386 architecture is deprecated for macOS (remove from the Xcode build setting: ARCHS)`

here is the whole output:

```
++++ info: making comb.rev~.o in lib echo
cc -DPD -I "/Applications/Pd.app/Contents/Resources/src" -DUNIX -DMACOSX -I /sw/include   -Ishared -DHAVE_STRUCT_TIMESPEC -Wall -Wextra -Wshadow -Winline -Wstrict-aliasing -Wno-unused-function -Wno-unused-parameter -Wno-unused-value -Wno-unused-variable -O3 -ffast-math -funroll-loops -fomit-frame-pointer -arch i386 -arch x86_64 -mmacosx-version-min=10.6 -o comb.rev~.o -c comb.rev~.c
++++ info: linking objects in comb.rev~.pd_darwin for lib echo
cc -undefined suppress -flat_namespace -bundle  -arch i386 -arch x86_64 -mmacosx-version-min=10.6  -o comb.rev~.pd_darwin comb.rev~.o  -lc   
ld: warning: The i386 architecture is deprecated for macOS (remove from the Xcode build setting: ARCHS)
ld: warning: ignoring file /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/usr/lib/libc.tbd, missing required architecture i386 in file /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/usr/lib/libc.tbd
ld: warning: ignoring file /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/usr/lib/libSystem.tbd, missing required architecture i386 in file /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/usr/lib/libSystem.tbd
```

**Comment by megrimm** [comment]
also... here is something else...

i tried 1137cca with another lib (flext based) and got this:

```
warning: include path for stdlibc++ headers not found; pass '-std=libc++' on the command line to use the libc++ standard library instead [-Wstdlibcxx-not-found]
In file included from clarinet~.cpp:3:
In file included from /usr/local/include/flext/flstk.h:12:
In file included from /usr/local/include/flext/flext.h:60:
In file included from /usr/local/include/flext/flclass.h:18:
In file included from /usr/local/include/flext/flbase.h:19:
/usr/local/include/flext/flsupport.h:17:10: fatal error: 'new' file not found
#include <new>
```
Looks like sometimes -mmacosx-version-min needs to be set to 10.9. im not exactly clear why but i did set this on line 581 and the build went proper.
see here: https://github.com/pandas-dev/pandas/issues/23424

maybe there is an option to pass -mmacosx-version-min from Makefile to pd-lib-builder? or maybe there already is that I do not know of?

thanks!

**Comment by katjav** [comment]
If a library needs a particular minimum OSX version it can be written in the Makefile like so:

define forDarwin
cflags = -mmacosx-version-min=10.9
endef

@megrimm, doesn't flext have its own build system, how can you combine this with pdlibbuilder? Can you point to the project? I wouldn't like to specify 10.9 as default minimum for all Pd libs.

**Comment by megrimm** [comment]
>> can be written in the Makefile like so

got it.. thats right thanks


>> doesn't flext have its own build system, how can you combine this with pdlibbuilder? Can you point to the project? 

i honestly can not seem to find where we borrowed this from but I think the idea was to bypass flext build system in order to create something more standardized with pd-lib-builder (one build system for all)

what i found is this:

```
## pd-lib-builder based Makefile for Flext externals
#
# instead of calling '.../flext/build.sh pd gcc'
# just do 'make -f .../Makefile.flext'
ifndef PKGINFO
PKGINFO=package.txt
endif
ifneq ($(PKGINFO),)
include $(PKGINFO)
endif
ifdef BUILDDIR
-include $(BUILDDIR)/config-lnx.def
-include $(BUILDDIR)/gnumake-lnx-gcc.inc
endif

lib.name = $(NAME)

$(NAME).class.sources = $(addprefix $(SRCDIR)/, $(SRCS))

cflags = $(INCPATH) $(DEFS) $(shell pkg-config --cflags pd-flext)
ldflags = $(shell pkg-config --libs pd-flext)

# This Makefile is based on the Makefile from pd-lib-builder written by
# Katja Vetter. You can get it from:
# https://github.com/pure-data/pd-lib-builder

PDLIBBUILDER_DIR=/usr/share/pd-lib-builder/
include $(firstword $(wildcard $(PDLIBBUILDER_DIR)/Makefile.pdlibbuilder Makefile.pdlibbuilder))
```

**Comment by katjav** [comment]
So it seems there's a quite a few things deprecated / discontinued. The current issue about not finding 10.5 is considered resolved, as 10.6 is OK for C code. For C++ there seems to be a different complication. And the deprecation of i386 also needs further investigation. Let's discuss these in seperate issues so that other people can find them under appropriate titles.

**Comment by umlaeute** [comment]
the pd-lib-builder for flext is [my feeble attempt](https://salsa.debian.org/multimedia-team/pd/pd-flext/blob/master/debian/Makefile.flext) to have flext-based externals in Debian.

I find flext-build system to be practically unusable for automated builds, and pd-lib-builder provided an easy alternative that would actually work.

**Comment by umlaeute** [comment]
@megrimm please be aware that the current "pd-lib-builder for flext" is not targetted at macOS builds¹.
specifically, it includes the flext-Makefile-snippets for **Linux** only, and relies on a pkg-config enabled installation of flext.

¹ this might change, but (imho) is out-of-scope for the pd-lib-builder project.

**Comment by megrimm** [comment]
>> flext-build system to be practically unusable for automated builds, and pd-lib-builder provided an easy alternative

maybe we can add a flext pd-lib-builder example. did the one you wrote @umlaeute ever work for you?

 >> "pd-lib-builder for flext" is not targetted at macOS builds

ah yes the one i sent in this thread was the unaltered one that must have come from you

maybe this is a separate issue because i think a cross-platform way to build flext externals using pd-lib-builder would be useful to simplify and standardize.

**Comment by umlaeute** [comment]
> maybe we can add a flext pd-lib-builder example.

i don't know. you would have to check with the flext build system.

> did the one you wrote @umlaeute ever work for you?

sure. why do you think otherwise? (i find the question mildly insulting)
as said above, "pd-lib-builder for flext" was created to include flext-based externals in Debian and derivatives.
at least [`pd-py`](https://tracker.debian.org/pkg/pd-py), [`pd-pool`](https://tracker.debian.org/pkg/pd-pool) and [`pd-xsample`](https://tracker.debian.org/pkg/pd-xsample) are built regularly with it. so far i haven't heard of *any* problems.

**Comment by katjav** [comment]
Was fixed with commit 1137cca2 which is part of v0.6.0.
